File: app/src/main/java/com/termirust/mobile/controller/ControllerRemoteRouteConfiguration.kt
package com.termirust.mobile.controller

import com.termirust.mobile.security.MobileSecretStore
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.net.URI

@Serializable
enum class ControllerRouteCredentialPurpose {
    @SerialName("ssh_authentication") SSH_AUTHENTICATION,
    @SerialName("relay_admission") RELAY_ADMISSION,
}

@Serializable
data class ControllerRouteCredentialReference(
    val id: String,
    val route: ControllerRemoteRouteKind,
    val purpose: ControllerRouteCredentialPurpose,
) {
    init {
        require(id.isSafeRouteIdentifier())
        require(
            (route == ControllerRemoteRouteKind.SSH && purpose == ControllerRouteCredentialPurpose.SSH_AUTHENTICATION) ||
                (route == ControllerRemoteRouteKind.SELF_HOSTED_RELAY && purpose == ControllerRouteCredentialPurpose.RELAY_ADMISSION),
        )
    }
}

@Serializable
data class ControllerRemoteRouteConfiguration(
    val kind: ControllerRemoteRouteKind,
    val endpoint: String,
    val port: Int? = null,
    val username: String? = null,
    @SerialName("trust_pin") val trustPin: String? = null,
    val credential: ControllerRouteCredentialReference? = null,
) {
    fun validate() {
        require(endpoint.isNotBlank() && endpoint.toByteArray().size <= 2_048 && endpoint.none(Char::isISOControl))
        when (kind) {
            ControllerRemoteRouteKind.LOCAL_IPC -> error("local IPC is not an Android route configuration")
            ControllerRemoteRouteKind.PRIVATE_NETWORK -> {
                require(port in 1..65_535 && username == null && trustPin == null && credential == null)
                HostRoute(endpoint, port!!)
            }
            ControllerRemoteRouteKind.SSH -> {
                require(port in 1..65_535)
                require(!username.isNullOrBlank() && username.toByteArray().size <= 255 && username.none(Char::isISOControl))
                require(!trustPin.isNullOrBlank() && trustPin.toByteArray().size <= 512 && trustPin.none(Char::isISOControl))
                require(credential?.route == kind && credential.purpose == ControllerRouteCredentialPurpose.SSH_AUTHENTICATION)
                HostRoute(endpoint, port!!)
            }
            ControllerRemoteRouteKind.SELF_HOSTED_RELAY -> {
                require(port == null && username == null)
                require(!trustPin.isNullOrBlank() && trustPin.toByteArray().size <= 512 && trustPin.none(Char::isISOControl))
                require(credential?.route == kind && credential.purpose == ControllerRouteCredentialPurpose.RELAY_ADMISSION)
                val uri = runCatching { URI(endpoint) }.getOrElse { throw IllegalArgumentException("invalid relay endpoint", it) }
                require(uri.scheme.equals("wss", ignoreCase = true) && !uri.host.isNullOrBlank())
                require(uri.userInfo == null && uri.fragment == null)
            }
        }
    }
}

class ControllerRouteCredentialStore(private val secrets: MobileSecretStore) {
    fun save(hostId: String, reference: ControllerRouteCredentialReference, secret: String) {
        require(secret.isNotEmpty() && secret.toByteArray().size <= 64 * 1_024)
        secrets.saveSecret(account(hostId, reference), secret)
    }

    fun read(hostId: String, reference: ControllerRouteCredentialReference): String? =
        secrets.readSecret(account(hostId, reference))

    fun delete(hostId: String, reference: ControllerRouteCredentialReference) {
        secrets.deleteSecret(account(hostId, reference))
    }

    internal fun account(hostId: String, reference: ControllerRouteCredentialReference): String {
        require(hostId.isSafeRouteIdentifier())
        return listOf(
            "controller-route-v1",
            hostId,
            reference.route.name.lowercase(),
            reference.purpose.name.lowercase(),
            reference.id,
        ).joinToString(":")
    }
}

private fun String.isSafeRouteIdentifier() =
    isNotBlank() && toByteArray().size <= 128 && all { it.isLetterOrDigit() || it in "-_." }

File: TermiRustMobile/Models/ControllerRemoteRouteConfiguration.swift
import Foundation

enum ControllerRouteCredentialPurpose: String, Codable, Sendable {
    case sshAuthentication = "ssh_authentication"
    case relayAdmission = "relay_admission"
}

struct ControllerRouteCredentialReference: Codable, Equatable, Hashable, Sendable {
    let id: String
    let route: ControllerRemoteRouteKind
    let purpose: ControllerRouteCredentialPurpose

    init(
        id: String,
        route: ControllerRemoteRouteKind,
        purpose: ControllerRouteCredentialPurpose
    ) throws {
        let trimmed = id.trimmingCharacters(in: .whitespacesAndNewlines)
        guard route != .localIPC,
              !trimmed.isEmpty,
              trimmed.utf8.count <= 128,
              !trimmed.unicodeScalars.contains(where: CharacterSet.controlCharacters.contains),
              (route == .ssh && purpose == .sshAuthentication)
                || (route == .selfHostedRelay && purpose == .relayAdmission) else {
            throw ControllerRemoteRouteConfigurationError.invalidCredentialReference
        }
        self.id = trimmed
        self.route = route
        self.purpose = purpose
    }
}

struct ControllerRemoteRouteConfiguration: Codable, Equatable, Sendable {
    let kind: ControllerRemoteRouteKind
    let endpoint: String
    let port: UInt16?
    let username: String?
    let trustPin: String?
    let credential: ControllerRouteCredentialReference?

    static func privateNetwork(endpoint: HostRoute) -> Self {
        Self(
            kind: .privateNetwork,
            endpoint: endpoint.address,
            port: endpoint.port,
            username: nil,
            trustPin: nil,
            credential: nil
        )
    }

    static func ssh(
        endpoint: String,
        port: UInt16,
        username: String,
        hostKeyPin: String,
        credential: ControllerRouteCredentialReference
    ) throws -> Self {
        let value = Self(
            kind: .ssh,
            endpoint: endpoint,
            port: port,
            username: username,
            trustPin: hostKeyPin,
            credential: credential
        )
        try value.validate()
        return value
    }

    static func selfHostedRelay(
        endpoint: String,
        spkiPin: String,
        credential: ControllerRouteCredentialReference
    ) throws -> Self {
        let value = Self(
            kind: .selfHostedRelay,
            endpoint: endpoint,
            port: nil,
            username: nil,
            trustPin: spkiPin,
            credential: credential
        )
        try value.validate()
        return value
    }

    func validate() throws {
        guard !endpoint.isEmpty, endpoint.utf8.count <= 2_048 else {
            throw ControllerRemoteRouteConfigurationError.invalidEndpoint
        }
        switch kind {
        case .localIPC:
            throw ControllerRemoteRouteConfigurationError.unsupportedRoute
        case .privateNetwork:
            guard let port, username == nil, trustPin == nil, credential == nil else {
                throw ControllerRemoteRouteConfigurationError.invalidCombination
            }
            _ = try HostRoute(address: endpoint, port: port)
        case .ssh:
            guard let port,
                  let username,
                  let trustPin,
                  let credential,
                  credential.route == .ssh,
                  credential.purpose == .sshAuthentication,
                  !username.isEmpty,
                  username.utf8.count <= 255,
                  !trustPin.isEmpty,
                  trustPin.utf8.count <= 512 else {
                throw ControllerRemoteRouteConfigurationError.invalidCombination
            }
            _ = try HostRoute(address: endpoint, port: port)
        case .selfHostedRelay:
            guard port == nil,
                  username == nil,
                  let trustPin,
                  let credential,
                  credential.route == .selfHostedRelay,
                  credential.purpose == .relayAdmission,
                  trustPin.utf8.count <= 512,
                  let components = URLComponents(string: endpoint),
                  components.scheme?.lowercased() == "wss",
                  components.host != nil,
                  components.user == nil,
                  components.password == nil,
                  components.fragment == nil else {
                throw ControllerRemoteRouteConfigurationError.invalidCombination
            }
        }
    }
}

enum ControllerRemoteRouteConfigurationError: Error, Equatable, Sendable {
    case unsupportedRoute
    case invalidEndpoint
    case invalidCredentialReference
    case invalidCombination
}
